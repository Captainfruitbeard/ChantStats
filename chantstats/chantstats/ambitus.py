from enum import Enum
from music21.interval import Interval

__all__ = ["get_ambitus", "AmbitusType"]


class AmbitusType(str, Enum):
    """
    Possible ambitus types for pieces.
    """

    AUTHENTIC = "authentic"
    PLAGAL = "plagal"
    UNDEFINED = "undefined"


def get_ambitus(*, note_of_final, lowest_note):
    if note_of_final is None:
        return AmbitusType.UNDEFINED

    interval = Interval(note_of_final, lowest_note)
    if 0 >= interval.semitones >= -4:
        return AmbitusType.AUTHENTIC
    elif -5 >= interval.semitones > -12:
        return AmbitusType.PLAGAL
    else:
        raise Exception(  # pragma: no cover
            "Check the logic in the ambitus calculation! "
            "We expect the lowest note to be less than an octave "
            "below the main final."
        )
