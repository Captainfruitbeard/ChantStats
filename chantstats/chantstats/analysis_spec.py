import os
import shutil
from abc import ABCMeta
from enum import Enum

from .dendrogram2 import Dendrogram
from .logging import logger
from .modal_category import ModalCategory, GroupingByModalCategory

__all__ = ["FullAnalysisSpec"]


class EnumWithDescription(str, Enum):
    def __new__(cls, name, desc, **kwargs):
        obj = str.__new__(cls, name)
        obj._value_ = name
        obj._description = desc
        return obj

    @property
    def description(self) -> str:
        """
        Returns
        -------
        str
            Long form description of this enum suitable for use in plot titles etc.
        """
        return self._description


class RepertoireAndGenreType(EnumWithDescription):
    PLAINCHANT_SEQUENCES = ("plainchant_sequences", "Plainchant Sequences", ("chant", "sequences"))
    RESPONSORIAL_CHANTS = ("responsorial_chants", "Responsorial Chants", ("chant", "responsorial_chants"))
    ORGANA = ("organa", "Organa", ("organum", ""))

    def __new__(cls, name, desc, path_stubs, **kwargs):
        obj = str.__new__(cls)
        obj._value_ = name
        obj._description = desc
        obj.path_stub_1 = path_stubs[0]
        obj.path_stub_2 = path_stubs[1]
        return obj


class BaseAnalysisFunc(metaclass=ABCMeta):
    def __call__(self, item):
        raise NotImplementedError()


class AnalysisFuncPCFreqs(BaseAnalysisFunc):
    def __init__(self, abs_or_rel):
        assert abs_or_rel in ["abs_freqs", "rel_freqs"]
        self.abs_or_rel = abs_or_rel

    def __call__(self, item, *, unit):
        if unit == UnitType.PCS:
            return getattr(item.pc_freqs, self.abs_or_rel)
        elif unit == UnitType.MODE_DEGREES:
            return getattr(item.mode_degree_freqs, self.abs_or_rel)
        else:
            raise RuntimeError("Unexpected unit: {self.unit}")


class AnalysisType(EnumWithDescription):
    PC_FREQS = ("pc_freqs", "PC Frequencies")
    TENDENCY = ("tendency", "Tendency")
    APPROACHES_AND_DEPARTURES = ("approaches_and_departures", "Approaches & Departures")
    LEAPS_AND_MELODIC_OUTLINES = ("leaps_and_melodic_outlines", "Leaps & Melodic Outlines")

    @property
    def analysis_func(self):
        if self.value == "pc_freqs":
            return AnalysisFuncPCFreqs("rel_freqs")
        else:
            return BaseAnalysisFunc()


class UnitType(EnumWithDescription):
    PCS = ("pcs", "pitch classes")
    MODE_DEGREES = ("mode_degrees", "mode degrees")


class ModeType(EnumWithDescription):
    BY_FINAL = ("by_final", "by final")
    AUTHENTIC_MODES = ("authentic_modes", "authentic modes")
    PLAGAL_MODES = ("plagal_modes", "plagal modes")

    def get_description(self, *, final):
        # TODO: check that final is a valid pitch class
        if self.value == "by_final":
            return f"{final}-final"
        elif self.value == "authentic_modes":
            return f"{final}-authentic"
        elif self.value == "plagal_modes":
            return f"{final}-plagal"
        else:
            raise RuntimeError(f"Unexpected value: {self.value!r}")

    def get_subfolder(self, *, final):
        # TODO: check that final is a valid pitch class
        if self.value == "by_final":
            return f"{final}_final"
        elif self.value == "authentic_modes":
            return f"{final}_authentic"
        elif self.value == "plagal_modes":
            return f"{final}_plagal"
        else:
            raise RuntimeError(f"Unexpected value: {self.value!r}")


class FullAnalysisSpecOLD:
    def __init__(self, *, repertoire_and_genre, analysis, unit, mode):
        self.repertoire_and_genre = RepertoireAndGenreType(repertoire_and_genre)
        self.analysis = AnalysisType(analysis)
        self.unit = UnitType(unit)
        self.mode = ModeType(mode)

        self.analysis_func = self.analysis.analysis_func

    def get_description(self, *, final):
        # TODO: assert that `final` is a valid pitch class
        return f"{self.analysis.description}, {self.repertoire_and_genre.description}, {self.mode.get_description(final=final)}"

    def get_description_NEW(self, *, modal_category):
        assert isinstance(modal_category, ModalCategory)
        return f"{self.analysis.description}, {self.repertoire_and_genre.description}, {modal_category.descr}"

    def __repr__(self):
        return (
            f"<FullAnalysisSpecOLD: "
            f"repertoire_and_genre={self.repertoire_and_genre.value!r}, "
            f"analysis={self.analysis.value!r}, "
            f"unit={self.unit.value!r}, "
            f"mode={self.mode.value!r}, "
            ">"
        )

    def output_path(self, *, root_dir, final):
        return os.path.join(
            root_dir,
            self.repertoire_and_genre.path_stub_1,
            self.analysis.value,
            self.repertoire_and_genre.path_stub_2,
            self.unit.value,
            self.mode.value,
            self.mode.get_subfolder(final=final),
        )

    def export_results(self, *, output_root_dir, grouping, p_cutoff, overwrite=False, sort_freqs_ascending=False):
        raise DeprecationWarning(
            "This method has been moved to the class GroupingByModalCategory. "
            "You should call grouping.export_results(analysis_spec=..., ...)."
        )


class FullAnalysisSpec:
    def __init__(self, *, repertoire_and_genre, analysis, unit):
        self.repertoire_and_genre = RepertoireAndGenreType(repertoire_and_genre)
        self.analysis = AnalysisType(analysis)
        self.unit = UnitType(unit)
        # self.mode = ModeType(mode)

        self.analysis_func = self.analysis.analysis_func

    def get_description(self, *, modal_category):
        assert isinstance(modal_category, ModalCategory)
        return f"{self.analysis.description}, {self.repertoire_and_genre.description}, {modal_category.descr}"

    def __repr__(self):
        return (
            f"<FullAnalysisSpec: "
            f"repertoire_and_genre={self.repertoire_and_genre.value!r}, "
            f"analysis={self.analysis.value!r}, "
            f"unit={self.unit.value!r}, "
            ">"
        )

    def output_path(self, *, root_dir, modal_category):
        return os.path.join(
            root_dir,
            self.repertoire_and_genre.path_stub_1,
            self.analysis.value,
            self.repertoire_and_genre.path_stub_2,
            self.unit.value,
            modal_category.output_path_stub,
        )

    def export_results(self, *, output_root_dir, grouping, p_cutoff, overwrite=False, sort_freqs_ascending=False):
        raise DeprecationWarning(
            "This method has been moved to the class GroupingByModalCategory. "
            "You should call grouping.export_results(analysis_spec=..., ...)."
        )
