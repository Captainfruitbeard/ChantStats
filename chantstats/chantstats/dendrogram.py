import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import os
import scipy.stats
from scipy.cluster.hierarchy import dendrogram, linkage
from scipy.spatial.distance import pdist

from .logging import logger

__all__ = ["plot_dendrogram"]


def calculate_chi_square_p_value(A):
    """
    Calulate the p-value of the Pearson chi-square test when applied
    to the rows of the matrix A (interpreted as individual observations)

    Note: any values where both freqs1 and freqs2 are zero are discarded
    before calculating the chi-square value.
    """
    nonzero_columns = np.where(A.any(axis=0))[0]
    A_nonzero_columns = A[:, nonzero_columns]
    _, p_value, _, _ = scipy.stats.chi2_contingency(A_nonzero_columns)
    return p_value


def calculate_distribution_distance(freqs1, freqs2):
    """
    Calulate the "distance" between two frequency distributions.
    This is (1-p), where p is the p-value of the chi-square calculation
    returned by the function `calculate_chi_square_p_value`.
    """
    A = np.array([freqs1, freqs2])
    p_value = calculate_chi_square_p_value(A)
    return 1 - p_value


class DendrogramError(Exception):
    """
    Indicates an error in the dendrogram calculation.
    """


def calculate_linkage_matrix_in_python_format(df_freq_distributions, *, optimal_ordering=True):
    if len(df_freq_distributions) <= 1:
        raise DendrogramError("Cannot produce dendrogram for a single item (nothing to cluster).")
    Z = linkage(
        pdist(df_freq_distributions.values, metric=calculate_distribution_distance),
        method="complete",
        optimal_ordering=optimal_ordering,
    )
    return Z


def plot_dendrogram(
    df_freq_distributions, *, p_threshold=0.15, figsize=(16, 5), leaf_font_size=10, optimal_ordering=True
):
    """
    Plot a dendrogram

    Parameters
    ----------
    df_freq_distributions : pandas.DataFrame
        Data frame containing the frequency distributions for which to calculate the dendrogram.
        Must contain one frequency distribution per row. The row labels will be used as the
        labels for the dendrogram leaf nodes.
    p_threshold : float
        Value of p which is considered the threshold for distributions to be considered "similar".
        All links above this threshold will be colored gray.
    figsize : pair of float
        Size of the output figure (in inches). Same as matplotlib figsize.
    leaf_font_size : int
        Font size to use for the labels of leaf notes.
    optimal_ordering : bool
        If True, the linkage matrix will be reordered so that the distance between successive leaves
        is minimal. This results in a more intuitive tree structure when the data are visualized.
        Note that this algorithm can be slow, particularly on large datasets. See the function
        `scipy.cluster.hierarchy.linkage` for details.

    Returns
    -------
    matplotlib.figure.Figure
        Matplotlib figure containing the dendrogram plot.
    """
    assert isinstance(df_freq_distributions, pd.DataFrame)
    Z = calculate_linkage_matrix_in_python_format(df_freq_distributions, optimal_ordering=optimal_ordering)
    labels = df_freq_distributions.index
    fig, ax = plt.subplots(figsize=figsize)
    ax.set_title("Hierarchical Clustering Dendrogram")
    # ax.set_xlabel('Pieces')
    ax.set_ylabel("Height")
    dendrogram(
        Z,
        labels=labels,
        truncate_mode=None,
        color_threshold=p_threshold,
        above_threshold_color="#999999",
        leaf_rotation=90.0,  # rotates the x axis labels
        leaf_font_size=leaf_font_size,  # font size for the x axis labels
        ax=ax,
    )
    ax.axhline(y=p_threshold, c="grey", lw=1, linestyle="dashed")
    ax.set_ylim(-0.05, 1.05)
    fig.tight_layout()
    plt.close(fig)
    return fig


class DendrogramNode:
    def __init__(self, *, descr, index, p_coeff, y_pos, payload, children):
        self.index = index
        self.p_coeff = p_coeff
        self.y_pos = y_pos
        self.payload = payload
        self.children = children
        self.parent = None

        self.is_leaf = self.children == []
        self.num_leaves = 1 if self.is_leaf else sum([c.num_leaves for c in self.children])
        self.num_descendants = sum([c.num_descendants for c in self.children]) + 1
        self.descr = descr or "Cluster with {} leaves".format(self.num_leaves)

        if payload is not None:
            assert isinstance(payload, pd.Series)
            self.payload = payload  # TODO: if payload==None, set this to the averaged distribution of all leaves
        else:
            # Set payload to the average of the leaf payloads
            leaf_payloads = [c.payload for c in self.leaves]
            zero_payload = pd.Series(0, index=leaf_payloads[0].index)
            self.payload = sum(leaf_payloads, zero_payload) / len(leaf_payloads)

        self.df = pd.DataFrame([n.payload for n in self.leaves])

    def __repr__(self):
        if self.is_leaf:
            descr_descendants = "leaf node"
        else:
            descr_descendants = (
                f"{self.num_descendants} descendants, "
                f"{self.num_leaves} leaves, "
                f"immediate children: {[c.index for c in self.children]}"
            )
        return f"<DendrogramNode: '{self.descr}', idx: {self.index} ({descr_descendants})>"

    @property
    def descendants(self):
        yield self
        for c in self.children:
            yield from c.descendants

    @property
    def leaves(self):
        for n in self.descendants:
            if n.is_leaf:
                yield n

    @property
    def left_child(self):
        try:
            return self.children[0]
        except IndexError:
            raise DendrogramError("Leaf node does not have any children.")

    @property
    def right_child(self):
        try:
            return self.children[1]
        except IndexError:
            raise DendrogramError("Leaf node does not have any children.")

    def get_max_nodes_below_cutoff(self, p_cutoff, *, exclude_leaf_nodes=True):
        if p_cutoff >= 1.0:
            raise ValueError("The value of p_cutoff must be less than 1.0")

        max_nodes = [n for n in self.descendants if n.p_coeff <= p_cutoff and n.parent.p_coeff > p_cutoff]
        if exclude_leaf_nodes:
            max_nodes = [n for n in max_nodes if not n.is_leaf]
        return max_nodes


def make_dendrogram_tree(df):
    """
    Create a dendrogram tree from frequency distributions.

    Parameters
    ----------
    df : pandas.DataFrame
        The input data frame. Must contain one row per distribution.

    Returns
    -------
    DendrogramNode
        The root node of the dendrogram.
    """
    ZZZ = calculate_linkage_matrix_in_python_format(df)

    # set idx to the index of the root node (= the one with the largest number of descendants)
    root_node_idx = int(ZZZ[:, 3].argmax())

    N = len(df)
    return make_dendrogram_subtree(df, root_node_idx + N, ZZZ, N)


def make_dendrogram_subtree(df, idx, ZZZ, N):

    if idx < N:
        # leaf node
        result = DendrogramNode(
            descr=str(df.index[idx]), index=idx, p_coeff=0, y_pos=100.0, payload=df.iloc[idx], children=[]
        )
    else:
        i = int(ZZZ[idx - N, 0])
        j = int(ZZZ[idx - N, 1])
        p_coeff = ZZZ[idx - N, 2]
        left_child = make_dendrogram_subtree(df, i, ZZZ, N)
        right_child = make_dendrogram_subtree(df, j, ZZZ, N)
        children = [left_child, right_child]
        result = DendrogramNode(
            descr=None, index=idx, p_coeff=p_coeff, y_pos=(100 * (1 - p_coeff)), payload=None, children=children
        )
        left_child.parent = result
        right_child.parent = result

    return result


def plot_bar_chart_for_dendrogram_node_payload(node, ymax=None, figsize=(8, 4)):
    """
    Plot a bar chart based on the payload of the given dendrogram node.

    Parameters
    ----------
    node : DendrogramNode
        The dendrogram node of which to plot the payload.
    ymax : float, optional
        Set the y-axis limit of the bar chart to this value (optional).
    figsize : (float, float), optional
        Width and height of the resulting figure in inches.

    Returns
    -------
    matplotlib.figure.Figure
    """
    assert isinstance(node, DendrogramNode)
    fig, ax = plt.subplots(figsize=figsize)
    node.payload.plot.bar(ax=ax)
    title = f"DendrogramNode with {len(list(node.leaves))} leaves (index: {node.index})"
    ax.set_title(title)
    ax.set_ylim(0, ymax)
    plt.close(fig)
    return fig


def export_max_nodes_below_cutoff(tree, *, output_dir, p_cutoff, ymax=None, fmt="pdf"):
    """
    Given a dendrogram tree, find the "maximal" nodes whose p-value
    is below `p_cutoff` and export a bar chart for each of these nodes'
    distributions in the directory `output_dir`.
    """
    assert isinstance(tree, DendrogramNode)

    supported_output_formats = ["pdf", "png"]
    if fmt not in supported_output_formats:
        raise ValueError(f"Unsupported output format: '{fmt}'. Supported formats are: {supported_output_formats}")

    ymax = ymax or tree.df.max().max()

    logger.debug(f"Exporting bar charts to directory: '{output_dir}'")
    os.makedirs(output_dir, exist_ok=True)
    nodes = tree.get_max_nodes_below_cutoff(p_cutoff=p_cutoff)
    for node in nodes:
        fig = plot_bar_chart_for_dendrogram_node_payload(node, ymax=ymax)
        outfilename = f"bar_chart_for_dendrogram_node_{node.index:03d}.{fmt}"
        logger.debug(f"Saving bar chart to file: '{outfilename}'")
        output_path = os.path.join(output_dir, outfilename)
        fig.savefig(output_path)
