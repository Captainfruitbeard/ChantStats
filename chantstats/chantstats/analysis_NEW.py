import os
import pickle
from collections import defaultdict
from tqdm import tqdm

from .analysis_spec import RepertoireAndGenreType, AnalysisType, AnalysisFuncPCFreqs
from .dendrogram2 import Dendrogram
from .logging import logger
from .modal_category import ModalCategoryType, GroupingByModalCategory
from .unit import UnitType
from .plainchant_sequence_monomodal_sections import extract_monomodal_sections

__all__ = ["prepare_analysis_inputs"]


def recursive_defaultdict():
    return defaultdict(recursive_defaultdict)


def get_recursive_subdict(the_dict, keys):
    first_key = keys[0]
    if len(keys) == 1:
        return the_dict[first_key]
    elif len(keys) > 1:
        return get_recursive_subdict(the_dict[first_key], keys[1:])
    else:
        raise RuntimeError()


def convert_to_nested_dict(d):
    if not isinstance(d, dict):
        return d
    else:
        return {key: convert_to_nested_dict(val) for key, val in d.items()}


class AnalysisResultCollection:
    def __init__(self):
        self._results = defaultdict(dict)
        self._results_nested = recursive_defaultdict()  # this is redudant (could compute on the fly), but nice to have

    def __getitem__(self, path_stubs):
        return self._results[path_stubs]

    def _get_path_stubs(self, repertoire_and_genre, analysis, unit, modal_category):
        rep_and_genre = RepertoireAndGenreType(repertoire_and_genre)
        analysis = AnalysisType(analysis)
        unit = UnitType(unit)
        return (
            rep_and_genre.output_path_stub_1,
            analysis.output_path_stub,
            rep_and_genre.output_path_stub_2,
            unit.output_path_stub,
            modal_category.output_path_stub_1,
            modal_category.output_path_stub_2,
        )

    def insert_results(self, repertoire_and_genre, analysis, unit, modal_category, results_key, value):
        path_stubs = self._get_path_stubs(repertoire_and_genre, analysis, unit, modal_category)
        get_recursive_subdict(self._results_nested, path_stubs)[results_key] = value
        self._results[path_stubs][results_key] = value

    def to_dict(self):
        return dict(self._results)  # convert from defaultdict to dict

    def to_nested_dict(self):
        return convert_to_nested_dict(self._results_nested)

    def to_pickle(self, filename, overwrite=False):
        if os.path.exists(filename):
            if not overwrite:
                logger.warn(f"Not overwriting existing file: '{filename}' (use 'overwrite=True' to overwrite)")
                return
            else:
                logger.warn(f"Overwriting existing file: '{filename}'")

        with open(filename, "wb") as f:
            pickle.dump(self, f)

    @classmethod
    def from_pickle(cls, filename):
        with open(filename, "rb") as f:
            return pickle.load(f)


def prepare_analysis_inputs(repertoire_and_genre, mode, *, cfg, min_length_monomodal_sections=3, filename_pattern=None):
    pieces = cfg.load_pieces(repertoire_and_genre, pattern=filename_pattern)
    mode = ModalCategoryType(mode)

    enforce_same_ambitus = {"final": False, "final_and_ambitus": True}[mode]

    if repertoire_and_genre == "plainchant_sequences":
        return extract_monomodal_sections(
            pieces, enforce_same_ambitus=enforce_same_ambitus, min_length=min_length_monomodal_sections
        )
    else:
        raise NotImplementedError(f"Unsupported repertoire/genre: {repertoire_and_genre}")


def accumulate_results(
    existing_results, cfg, repertoire_and_genre, analysis, unit, mode, min_length_monomodal_sections=3, p_cutoff=0.7
):
    results = existing_results or AnalysisResultCollection()
    analysis_inputs = prepare_analysis_inputs(
        repertoire_and_genre, mode, cfg=cfg, min_length_monomodal_sections=3, filename_pattern=None
    )

    grouping = GroupingByModalCategory(analysis_inputs, group_by=mode)
    logger.debug(f"Calculating results for {grouping}")
    for key in tqdm(grouping.keys):
        modal_category = grouping[key]
        df = modal_category.make_results_dataframe(analysis_func=AnalysisFuncPCFreqs("rel_freqs"), unit=unit)
        dendrogram = Dendrogram(df, p_threshold=p_cutoff)

        results.insert_results(repertoire_and_genre, analysis, unit, modal_category, "dendrogram", dendrogram.to_json())
        results.insert_results(
            repertoire_and_genre,
            analysis,
            unit,
            modal_category,
            "clusters",
            [c.to_json() for c in dendrogram.nodes_below_cutoff],
        )

    return results
