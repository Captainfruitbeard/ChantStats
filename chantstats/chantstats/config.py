from .analysis_spec import RepertoireAndGenreType


class ChantStatsConfig:
    def __init__(self, musicxml_paths):
        self.musicxml_paths = {RepertoireAndGenreType(key): path for (key, path) in musicxml_paths.items()}

    def get_musicxml_path(self, repertoire_and_genre):
        rep_and_genre = RepertoireAndGenreType(repertoire_and_genre)
        musicxml_path = self.musicxml_paths[rep_and_genre]
        return musicxml_path

    def load_pieces(self, repertoire_and_genre, *, pattern=None):
        raise DeprecationWarning("This method is deprecated. Use the function 'load_pieces()' instead.")
