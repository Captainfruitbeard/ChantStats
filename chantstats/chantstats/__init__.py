from .analysis_spec import FullAnalysisSpec
from .config import ChantStatsConfig
from .dendrogram import plot_dendrogram
from .logging import logger
from .modal_category import GroupingByModalCategory
from .plainchant_sequence_piece import load_plainchant_sequence_pieces, load_pieces
from .plainchant_sequence_monomodal_sections import MonomodalSection, extract_monomodal_sections
from .phrase_collection import PhraseCollection
from .results_export_OLD import export_dendrogram_and_stacked_bar_chart

__all__ = [
    "FullAnalysisSpec",
    "PhraseCollection",
    "GroupingByModalCategory",
    "ChantStatsConfig",
    "load_pieces",
    "load_plainchant_sequence_pieces",
    "logger",
    "chantstats_logger",
    "plot_dendrogram",
    "export_dendrogram_and_stacked_bar_chart",
]

chantstats_logger = logger  # alias for convenience

# Set __version__ attribute (via versioneer)
from ._version import get_versions

__version__ = get_versions()["version"]
del get_versions
