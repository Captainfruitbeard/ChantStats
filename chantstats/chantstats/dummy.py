from itertools import count

import numpy as np

from chantstats import GroupingByModalCategory
from chantstats.pitch_class_freqs import PCFreqs

pcs = ["A", "B", "C", "D", "E", "F", "G"]


class DummyItem:
    counter = count(start=1)

    def __init__(self, *, final, ambitus, index=None):
        self.index = index or next(self.counter)
        self.descr = f"DummyItem #{self.index}"
        self.final = final
        self.ambitus = ambitus
        self.random_state = np.random.RandomState()
        self.random_state.seed(self.index)
        self.pitch_classes = list(self.random_state.choice(pcs, size=100))
        self.pc_freqs = PCFreqs(self.pitch_classes)

    def __repr__(self):
        return f"<DummyItem #{self.index!r}: final={self.final!r}, ambitus={self.ambitus!r}>"


def make_dummy_items(*, final, ambitus, num):
    return [DummyItem(final=final, ambitus=ambitus) for _ in range(num)]


dummy_items = sum(
    [
        make_dummy_items(final="C", ambitus="authentic", num=5),
        make_dummy_items(final="D", ambitus="authentic", num=10),
        make_dummy_items(final="D", ambitus="plagal", num=12),
        make_dummy_items(final="G", ambitus="authentic", num=20),
        make_dummy_items(final="G", ambitus="plagal", num=18),
    ],
    [],
)


def make_dummy_grouping_by_modal_category(*, group_by):
    return GroupingByModalCategory(dummy_items, group_by=group_by)
