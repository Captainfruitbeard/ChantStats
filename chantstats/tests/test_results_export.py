import pytest

from .context import chantstats
from chantstats import FullAnalysisSpec


@pytest.mark.parametrize(
    "repertoire_and_genre, analysis, unit, mode, final, output_path_expected",
    [
        (
            "plainchant_sequences",
            "pc_freqs",
            "pcs",
            "by_final",
            "D",
            "/foobar/chant/pc_freqs/sequences/pcs/by_final/D_final",
        ),
        (
            "responsorial_chants",
            "tendency",
            "mode_degrees",
            "authentic_modes",
            "G",
            "/foobar/chant/tendency/responsorial_chants/mode_degrees/authentic_modes/G_authentic",
        ),
        (
            "organa",
            "approaches_and_departures",
            "pcs",
            "plagal_modes",
            "C",
            "/foobar/organum/approaches_and_departures/pcs/plagal_modes/C_plagal",
        ),
    ],
)
def test_output_path(repertoire_and_genre, analysis, unit, mode, final, output_path_expected):
    """
    Analysis spec returns the expected output path, depending on its input parameters.
    """
    spec = FullAnalysisSpec(repertoire_and_genre=repertoire_and_genre, analysis=analysis, unit=unit, mode=mode)
    output_path = spec.output_path(root_dir="/foobar", final=final)
    assert output_path_expected == output_path
