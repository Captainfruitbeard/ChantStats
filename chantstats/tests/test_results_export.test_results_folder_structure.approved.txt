.
└── chant
    └── pc_freqs
        └── sequences
            └── pcs
                └── authentic_modes
                    ├── A_authentic
                    │   ├── dendrogram.png
                    │   └── stacked_bar_chart.png
                    ├── B_authentic
                    │   ├── dendrogram.png
                    │   └── stacked_bar_chart.png
                    ├── C_authentic
                    │   ├── dendrogram.png
                    │   └── stacked_bar_chart.png
                    ├── D_authentic
                    │   ├── dendrogram.png
                    │   └── stacked_bar_chart.png
                    ├── E_authentic
                    │   ├── dendrogram.png
                    │   └── stacked_bar_chart.png
                    └── G_authentic
                        ├── dendrogram.png
                        └── stacked_bar_chart.png

11 directories, 12 files
